src/Features/Core/Portable/GenerateConstructorFromMembers/AbstractGenerateConstructorFromMembersCodeRefactoringProvider.GenerateConstructorWithDialogCodeAction.cs

```diff
@@ -66,10 +66,10 @@ public override object GetOptions(CancellationToken cancellationToken)
                     // the user chose.  That way we'll keep that as the default for the next time
                     // the user opens the dialog.
                     var workspace = _document.Project.Solution.Workspace;
-                    workspace.Options = workspace.Options.WithChangedOption(
+                    workspace.TryApplyChanges(workspace.CurrentSolution.WithOptions(workspace.Options.WithChangedOption(
```

Comment:
Tagging @tmat @jasonmalinowski @sharwell - The change made in this file to switch from using the Workspace Options setter to using `TryApplyChanges` is leading to integration test failure. This is due to the fact that this code is currently executing from a CodeAction which executes on a background thread. Switching to TryApplyChanges now throws an exception below: https://github.com/dotnet/roslyn/blob/b2ae91b793a98c90be5b585cb95858de3d707c00/src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs#L368-L375

What would be our recommended migration approach for this scenario?

Reply:
One possible approach that comes to mind is the code action should return a solution snapshot with changed option, instead of attempting to apply the option changes to the workspace itself. Does that sound reasonable?

Reply:
That seems to have worked: https://github.com/dotnet/roslyn/pull/40934/commits/e06510e96bdc1bcc04190aa5e34df4ceb8ec63a0.

Reply:
So I would have assumed this code was returning a CodeActionOperation that did the option change, and that would be running on the UI thread in the Operation's Apply. Is this actually being ran during background computation that then could be cancelled? I would have assumed this code had to be side-effect free.

Reply:
The option change was actually being done in `ComputeOperationsAsync` override, not the returned `CodeActionOperation`.

Reply:
Definitely. `ComputeOperationsAsync` should not access workspace.